optapy-core/src/main/python/test/__init__.py

```diff
@@ -0,0 +1,609 @@
+        return PythonSingleConstraintAssertion(self.delegate.givenSolution(wrapped_solution))
+
+
+class PythonMultiConstraintVerification(Generic[Solution_]):
+    def __init__(self, delegate):
+        self.delegate = delegate
+
+    def given(self, *facts) -> 'PythonMultiConstraintAssertion':
+        """
+        Set the facts for this assertion
+        :param facts: Never None, at least one
+        """
+        from org.optaplanner.optapy import PythonSolver  # noqa
+        from org.optaplanner.jpyinterpreter import CPythonBackedPythonInterpreter  # noqa
+        from org.optaplanner.jpyinterpreter.types import CPythonBackedPythonLikeObject  # noqa
+        from org.optaplanner.jpyinterpreter.types.wrappers import OpaquePythonReference  # noqa
+        reference_map = PythonSolver.getNewReferenceMap()
+        wrapped_facts = []
+
+        for fact in facts:
+            fact_class = get_class(type(fact))
+            wrapped_fact = PythonSolver.wrapFact(fact_class, fact, reference_map)
+            wrapped_facts.append(wrapped_fact)
+
+        for fact in reference_map.values():
+            if isinstance(fact, CPythonBackedPythonLikeObject):
+                CPythonBackedPythonInterpreter.updateJavaObjectFromPythonObject(fact, JProxy(OpaquePythonReference,
+                                                                                inst=getattr(fact, '$cpythonReference'),
+                                                                                convert=True), reference_map)
+
+        return PythonMultiConstraintAssertion(self.delegate.given(wrapped_facts))
+
+    def given_solution(self, solution: 'Solution_') -> 'PythonMultiConstraintAssertion':
+        """
+        Set the solution to be used for this assertion
+        :param solution: Never None
+        """
+        return self.givenSolution(solution)
+
+    def givenSolution(self, solution: 'Solution_') -> 'PythonMultiConstraintAssertion':
+        """
+        Set the solution to be used for this assertion
+        :param solution: Never None
+        """
+        from org.optaplanner.optapy import PythonSolver  # noqa
+        solution_class = get_class(type(solution))
+        wrapped_solution = PythonSolver.wrapProblem(solution_class, solution)
+        return PythonMultiConstraintAssertion(self.delegate.givenSolution(wrapped_solution))
+
+
+class PythonSingleConstraintAssertion:
+    def __init__(self, delegate):
+        self.delegate = delegate
+
+    @overload
+    def penalizes(self) -> None:
+        """
+        Asserts that the Constraint being tested, given a set of facts, results in any number of penalties.
+
+        Ignores the constraint and match weights: it only asserts the number of matches
+        For example: if there are two matches with weight of 10 each, this assertion will succeed.
+        If there are no matches, it will fail.
+
+        :raises AssertionError: when there are no penalties
+        """
+        ...
```

Comment:
I understand `...` to mean `pass`. That said, I'm not sure what this means in the context of a method. Is this how Python does abstract methods? Why would it be overloaded then?

Reply:
This is how Python declares type signatures for overloaded methods: have a bunch of functions that just do `pass` that declares the signature, followed by the function that implements all of them.